Sources/AWSLambdaDeploymentDescriptor/DeploymentDescriptorBuilder.swift

```diff
@@ -0,0 +1,424 @@
+        return DeploymentDescriptor(description: function.0, resources: function.1)
+    }
+    public static func buildExpression(_ expression: String) -> String {
+        return expression
+    }
+    public static func buildExpression(_ expression: Function) -> [Resource<ResourceType>] {
+        return expression.resources()
+    }
+    public static func buildExpression(_ expression: Queue) -> [Resource<ResourceType>] {
+        return [expression.resource()]
+    }
+    public static func buildExpression(_ expression: Table) -> [Resource<ResourceType>] {
+        return [expression.resource()]
+    }
+    public static func buildExpression(_ expression: Resource<ResourceType>) -> [Resource<ResourceType>] {
+        return [expression]
+    }
+}
+
+// MARK: Function resource
+public struct Function {
+    let name: String
+    let architecture: Architectures
+    let eventSources: [Resource<EventSourceType>]
+    let environment: [String: String]
+    
+    private init(
+        name: String,
+        architecture: Architectures = Architectures.defaultArchitecture(),
+        eventSources: [Resource<EventSourceType>] = [],
+        environment: [String: String] = [:]
+    ) {
+        self.name = name
+        self.architecture = architecture
+        self.eventSources = eventSources
+        self.environment = environment
+    }
+    public init(
+        name: String,
+        architecture: Architectures = Architectures.defaultArchitecture(),
+        @FunctionBuilder _ builder: () -> (EventSources, [String: String])
+    ) {
+        
+        let (eventSources, environmentVariables) = builder()
+        let samEventSource: [Resource<EventSourceType>] = eventSources.samEventSources()
+        self.init(
+            name: name,
+            architecture: architecture,
+            eventSources: samEventSource,
+            environment: environmentVariables)
+    }
+    
+    // this method fails when the package does not exist at path
+    internal func resources() -> [Resource<ResourceType>] {
+        
+        let functionResource = [
+            Resource<ResourceType>.serverlessFunction(
+                name: self.name,
+                architecture: self.architecture,
+                codeUri: packagePath(),
+                eventSources: self.eventSources,
+                environment: SAMEnvironmentVariable(self.environment))
+        ]
+        let additionalQueueResources = collectQueueResources()
+        
+        return functionResource + additionalQueueResources
+    }
+    
+    // compute the path for the lambda archive
+    private func packagePath() -> String {
+        
+        // propose a default path unless the --archive-path argument was used
+        // --archive-path argument value must match the value given to the archive plugin --output-path argument
+        var lambdaPackage =
+        ".build/plugins/AWSLambdaPackager/outputs/AWSLambdaPackager/\(self.name)/\(self.name).zip"
```

Comment:
this feels a bit too brittle. maybe there is a way to ask the plugin to construct this path for us?

Reply:
The plugin context should provide the path to the output directory already

Reply:
But this is another plugin.  This code is executed by the "deploy" plugin. The "deploy" plugin needs to figure out what was the default build path of the "archive" plugin

Reply:
I implemented the following strategy.
1. use the path provided on the plugin command line (--archive-path) 
2. otherwise use the path provided by the function developer in the deployment descriptor DSL 
3. otherwise fall back to the default hard coded to the default output of the archive plugin

Reply:
marking this as resolved, happy to dive deeper if you think it is necessary